src/packages/client/src/__tests__/integration/happy/long-running-transaction/test.ts

```diff
@@ -0,0 +1,374 @@
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+            Invalid \`prisma.user.create()\` invocation:
+
+
+              Unique constraint failed on the fields: (\`email\`)
+          `)
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(0)
+  })
+
+  /**
+   * A transaction that is already 'commit' cannot be reused
+   */
+  test('already committed', async () => {
+    let transactionBoundPrisma
+    prisma.$transaction((prisma) => {
+      transactionBoundPrisma = prisma
+    })
+
+    const result = prisma.$transaction(async () => {
+      await transactionBoundPrisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      })
+    })
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+            Invalid \`prisma.user.create()\` invocation:
+
+
+              Transaction API error: Transaction already closed: Transaction is no longer valid. Last state: 'Committed'.
+          `)
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(0)
+  })
+
+  /**
+   * Batching should work with using the interactive transaction logic
+   */
+  test('batching', async () => {
+    await prisma.$transaction([
+      prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      }),
+      prisma.user.create({
+        data: {
+          email: 'user_2@website.com',
+        },
+      }),
+    ])
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(2)
+  })
+
+  /**
+   * A bad batch should rollback using the interactive transaction logic
+   */
+  test('batching rollback', async () => {
+    const result = prisma.$transaction([
+      prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      }),
+      prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      }),
+    ])
+
+    await expect(result).rejects.toThrowErrorMatchingInlineSnapshot(`
+
+Invalid \`prisma.user.create()\` invocation:
+
+
+  Unique constraint failed on the fields: (\`email\`)
+`)
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(0)
+  })
+
+  /**
+   * Middlewares should work normally on batches
+   */
+  test('middlewares batching', async () => {
+    prisma.$use(async (params, next) => {
+      const result = await next(params)
+
+      return result
+    })
+
+    await prisma.$transaction([
+      prisma.user.create({
+        data: {
+          email: 'user_1@website.com',
+        },
+      }),
+      prisma.user.create({
+        data: {
+          email: 'user_2@website.com',
+        },
+      }),
+    ])
+
+    const users = await prisma.user.findMany()
+
+    expect(users.length).toBe(2)
+  })
+
+  /**
+   * Middlewares should not prevent a rollback
```

Comment:
Add test with middleware and bad input (aka not sent to the engine).

Reply:
Ok, added to my todo list.